pkg/networkservice/common/onidle/server.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2021 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package onidle provides server chain element that executes a callback when there were no connections for specified time
+package onidle
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/clock"
+)
+
+type onIdleServer struct {
+	ctx         context.Context
+	timeout     time.Duration
+	notify      func()
+	activeConns sync.Map
```

Comment:
If we already need a mutex to work with timer, probably it would be better to use `map[string]struct{}` instead of `sync.Map`, because there is `len()` method for the map and so there will be no need for `timerCounter`.